### ISSUE TITLE
setproctitle causes "Expected in: flat namespace" error

### ISSUE THREAD

**Issue by mamilov** [apache/airflow#8172]
**Apache Airflow version**: `1.10.9`

**Environment**:
Cloud provider or hardware configuration: Hardware: MacBook Pro 16-inch
OS (e.g. from /etc/os-release): MacOS 10.15.2
Kernel (e.g. uname -a): Darwin Kernel Version 19.2.0
Install tools: Pyenv & Pipenv
Others: None
Database: PostgreSQL

**What happened**:

I cannot import DAG into my project. When running `from airflow import DAG` I get an error.

**What you expected to happen**:

The expected behavior is that I'm able to import the DAG.  
The error seems to come from the setproctitle lib. Just importing that with ` import setproctitle`  throws the same error.

**How to reproduce it**:

Just import the DAG with `from airflow import DAG`.

**Anything else we need to know**:

How often does this problem occur? Every time

Any relevant logs to include? Put them here in side a detail tag:

<details><summary>airflow_import_dag_error.log</summary>Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/airflow/__init__.py", line 42, in <module>
    from airflow.models import DAG
  File "/Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/airflow/models/__init__.py", line 21, in <module>
    from airflow.models.baseoperator import BaseOperator, BaseOperatorLink  # noqa: F401
  File "/Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/airflow/models/baseoperator.py", line 43, in <module>
    from airflow.models.dag import DAG
  File "/Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/airflow/models/dag.py", line 46, in <module>
    from airflow.executors import LocalExecutor, get_default_executor
  File "/Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/airflow/executors/__init__.py", line 24, in <module>
    from airflow.executors.base_executor import BaseExecutor # noqa
  File "/Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/airflow/executors/base_executor.py", line 24, in <module>
    import airflow.utils.dag_processing
  File "/Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/airflow/utils/dag_processing.py", line 40, in <module>
    from setproctitle import setproctitle
ImportError: dlopen(/Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/setproctitle.cpython-37m-darwin.so, 2): Symbol not found: _Py_GetArgcArgv
  Referenced from: /Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/setproctitle.cpython-37m-darwin.so
  Expected in: flat namespace
 in /Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/setproctitle.cpython-37m-darwin.so</details>

<details><summary>setproctitle_import_error.log</summary>Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: dlopen(/Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/setproctitle.cpython-37m-darwin.so, 2): Symbol not found: _Py_GetArgcArgv
  Referenced from: /Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/setproctitle.cpython-37m-darwin.so
  Expected in: flat namespace
 in /Users/milov/.local/share/virtualenvs/dags-dwx8ai-u/lib/python3.7/site-packages/setproctitle.cpython-37m-darwin.so</details>

**Comment by mik-laj** [comment]
Related issue: https://github.com/apache/airflow/issues/8079

**Comment by ashb** [comment]
@mamilov Are you able to try a virtualenvironment outside of pyenv/pipenv please?

**Comment by jdavidheiser** [comment]
I also ran into this on Mac OS Catalina, but while running in a Bazel build environment that should theoretically have some isolation from system Python.

As a workaround, I was able to use Anaconda to install setproctitle and that worked.

**Comment by cboden** [comment]
I'm having this same problem. MacOS 10.15.3 with Python 3.7.6 installed via Homebrew installed inside a native venv (`python3 -m venv my-venv`).

**Comment by skylarbpayne** [comment]
## Potential Fix
Just see this to skip all the ugly details:

So concretely, here's how to install with homebrew + get airflow to work (tried this from fresh install, so I _think_ it should work for all of you too):

```
# Note: these instructions are for python3.7 but can be loosely modified for other versions
virtualenv -p /usr/local/opt/python@3.7/Frameworks/Python.framework/Versions/3.7/bin/python3 .toy-venv
source .toy-venv/bin/activate
pip install apache-airflow
python
>>> import setproctitle
# Success!
```

## Ugly Details

Mercurial ran into a similar issue some time ago: https://bz.mercurial-scm.org/show_bug.cgi?id=5858

According to that ticket, the issue is that setproctitle uses a non-public Python API, and depending on how your python was compiled, it may not be discoverable by setproctitle.

Versions of python installed to OS X with the standard python installer (i.e. downloaded from the python website) seem to always have a file like:

```
/Library/Frameworks/Python.framework/Versions/3.7/lib/libpython3.7.dylib
```

And if we use `nm` on this, we find that it contains the definition `setproctitle` relies on:
```
nm -gU /Library/Frameworks/Python.framework/Versions/3.7/lib/libpython3.7.dylib | grep -i argcargv
0000000000135630 T _Py_GetArgcArgv
```

By default, the "bin" directory used by homebrew is `/usr/local/opt/bin` -- libpython is not there. But there's also a `Frameworks` install (Note: I'm not smart enough to understand why this divide exists -- maybe someone knowledgeable about OS X, Python, and/or Homebrew can chime in :) ).

If we check out the Frameworks install, it has the dylib and has the required method defined:

```
nm -gU /usr/local/opt/python@3.7/Frameworks/Python.framework/Versions/3.7/lib/libpython3.7.dylib | grep -i argcargv
```

**Comment by ashb** [comment]
This sounds like nothing we can fix in airflow - perhaps just a doc/faq page?

Anyone fancy tithing this in to a PR?

**Comment by skylarbpayne** [comment]
@ashb -- happy to. I just linked a PR for this. LMK if anything needs to be added/edited/(re)moved

Thanks

**Comment by ashb** [comment]
Fixed by #12025
